File: apps/mobile/lib/api/imports.ts
import { apiRequest } from '@/lib/api/client';
import { EncodingType, readAsStringAsync } from 'expo-file-system/legacy';

export type CSVImportPreviewRow = {
  amount: string;
  category_id: string | null;
  category_name: string | null;
  fingerprint: string;
  note: string | null;
  raw_preview: Record<string, string>;
  row_index: number;
  title: string;
  transaction_date: string;
  type: 'expense' | 'income';
};

export type CSVImportSkippedRow = {
  raw_preview: Record<string, string>;
  reason: string;
  row_index: number;
};

export type CSVImportPreviewResponse = {
  detected_columns: string[];
  parsed_count: number;
  rows: CSVImportPreviewRow[];
  skipped_count: number;
  skipped_rows: CSVImportSkippedRow[];
  source_name: string | null;
};

export type CSVImportConfirmPayload = {
  original_parsed_count?: number | null;
  rows: Array<{
    amount: string;
    category_id: string | null;
    fingerprint: string;
    note: string | null;
    row_index: number;
    title: string;
    transaction_date: string;
    type: 'expense' | 'income';
  }>;
  source_name: string | null;
};

export type CSVImportConfirmResponse = {
  imported_count: number;
  imported_transaction_ids: string[];
  skipped_duplicate_count: number;
  skipped_duplicates: number[];
  source_name: string | null;
};

export type ImportHistoryItem = {
  id: string;
  source_name: string | null;
  original_parsed_count: number;
  requested_count: number;
  imported_count: number;
  ignored_count: number;
  skipped_duplicate_count: number;
  transaction_date_from: string | null;
  transaction_date_to: string | null;
  created_at: string;
  updated_at: string;
};

export type ImportHistoryResponse = {
  items: ImportHistoryItem[];
};

type UploadableCSVFile = {
  mimeType?: string | null;
  name: string;
  uri: string;
};

export async function previewCSVImport(
  accessToken: string,
  file: UploadableCSVFile,
): Promise<CSVImportPreviewResponse> {
  const extension = file.name.toLowerCase().split('.').pop();

  if (extension === 'xlsx') {
    const contentBase64 = await readAsStringAsync(file.uri, {
      encoding: EncodingType.Base64,
    });

    return apiRequest<CSVImportPreviewResponse>('/imports/xlsx/preview-base64', {
      accessToken,
      body: {
        content_base64: contentBase64,
        source_name: file.name,
      },
      method: 'POST',
    });
  }

  const content = await readAsStringAsync(file.uri);

  return apiRequest<CSVImportPreviewResponse>('/imports/csv/preview-text', {
    accessToken,
    body: {
      content,
      source_name: file.name,
    },
    method: 'POST',
  });
}

export function confirmCSVImport(accessToken: string, payload: CSVImportConfirmPayload) {
  return apiRequest<CSVImportConfirmResponse>('/imports/csv/confirm', {
    accessToken,
    body: payload,
    method: 'POST',
  });
}

export function getImportHistory(accessToken: string) {
  return apiRequest<ImportHistoryResponse>('/imports/history', {
    accessToken,
  });
}

File: apps/api/app/schemas/imports.py
from datetime import date
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import TransactionType


class CSVImportPreviewRow(BaseModel):
    row_index: int
    transaction_date: date
    title: str
    amount: Decimal
    type: TransactionType
    note: str | None = None
    category_id: UUID | None = None
    category_name: str | None = None
    fingerprint: str
    raw_preview: dict[str, str]


class CSVImportSkippedRow(BaseModel):
    row_index: int
    reason: str
    raw_preview: dict[str, str]


class CSVImportPreviewResponse(BaseModel):
    source_name: str | None = None
    detected_columns: list[str]
    parsed_count: int
    skipped_count: int
    rows: list[CSVImportPreviewRow]
    skipped_rows: list[CSVImportSkippedRow]


class CSVImportPreviewTextRequest(BaseModel):
    source_name: str | None = Field(default=None, max_length=255)
    content: str = Field(min_length=1)


class XLSXImportPreviewBase64Request(BaseModel):
    source_name: str | None = Field(default=None, max_length=255)
    content_base64: str = Field(min_length=1)


class CSVImportConfirmRow(BaseModel):
    row_index: int
    transaction_date: date
    title: str = Field(min_length=1, max_length=120)
    amount: Decimal = Field(gt=0)
    type: TransactionType
    note: str | None = None
    category_id: UUID | None = None
    fingerprint: str


class CSVImportConfirmRequest(BaseModel):
    source_name: str | None = Field(default=None, max_length=255)
    original_parsed_count: int | None = Field(default=None, ge=0)
    rows: list[CSVImportConfirmRow] = Field(min_length=1)


class CSVImportConfirmResponse(BaseModel):
    source_name: str | None = None
    imported_count: int
    skipped_duplicate_count: int
    skipped_duplicates: list[int]
    imported_transaction_ids: list[UUID]


class ImportBatchHistoryItem(BaseModel):
    id: UUID
    source_name: str | None = None
    original_parsed_count: int
    requested_count: int
    imported_count: int
    ignored_count: int
    skipped_duplicate_count: int
    transaction_date_from: date | None = None
    transaction_date_to: date | None = None
    created_at: str
    updated_at: str


class ImportBatchHistoryResponse(BaseModel):
    items: list[ImportBatchHistoryItem]
